src/Module.ts

```diff
@@ -119,6 +119,31 @@ const MISSING_EXPORT_SHIM_DESCRIPTION: ExportDescription = {
 	localName: MISSING_EXPORT_SHIM_VARIABLE
 };
 
+function findSourceMappingURLComments(ast: acorn.Node, code: string): [number, number][] {
+	let lastStmtEnd = 0;
+	const ranges = [];
+
+	for (const stmt of (ast as GenericEsTreeNode).body) {
+		if (lastStmtEnd != stmt.start) {
+			ranges.push([lastStmtEnd, stmt.start]);
+		}
+		lastStmtEnd = stmt.end;
+	}
+	if (lastStmtEnd != code.length) {
+		ranges.push([lastStmtEnd, code.length]);
+	}
+
+	const ret: [number, number][] = [];
+	for (const [start, end] of ranges) {
+		let sourcemappingUrlMatch;
+		while (sourcemappingUrlMatch = SOURCEMAPPING_URL_COMMENT_RE.exec(code.slice(start, end))) {
+			ret.push([start + sourcemappingUrlMatch.index, start + SOURCEMAPPING_URL_COMMENT_RE.lastIndex]);
+		}
+	}
+
+	return ret;
```

Comment:
I think there might be some potential to speed this up: Allocating arrays and slices of strings for every top-level statement is not for free as I can tell you from experience.

If you would do this the other way around, it might be much more efficient:
- Run your (stateful) regular expression over the entire code (in a loop). This should be quite fast as it is not creating a lot of new data structures (in the current edition, you are running it over all the code that is part of a top-level statement, which is nearly the same amount of code, but split up in many steps)
- In this loop for each match, determine if it is in a top-level statement. This might be done by just looping over the statements, or quite fast by doing a binary search through the statement list. I can recommend this algorithm as it does not use recursion and does not allocate any data structures: https://flaviocopes.com/binary-search-javascript/

Reply:
> in the current edition, you are running it over all the code that is part of a top-level statement

No, I am not. I am running over the **spaces between top-level statements**. Which shouldn't be very large - usually only newlines and comments.

> Allocating arrays and slices of strings for every top-level statement is not for free

AFAIK slices of strings in V8 are pretty much for free if it's RO (SlicedString is just a pointer to the parent and offset), see this https://mrale.ph/blog/2016/11/23/making-less-dart-faster.html for more details. And these slices, as mentioned, are quite small anyway (and they live very shortly, which means they die in the allocator's young space and never re-allocated).

The `ranges` array can't be overwhelmingly big, all on all the number of top-level statements is something humans should be able to cope with. So it can't be in the millions...

Do you still think it would be better to implement the idea you suggested?
If so, I will but I am afraid it will hurt readability (which IMHO is more important).

Reply:
> No, I am not. I am running over the spaces between top-level statements

Ah I see, I misread that. But you could still get rid of 

> The ranges array can't be overwhelmingly big
I would not underestimate this, if a file is not deeply nested it is not too far off from the order of the AST nodes (with a factor of maybe 10 between those)

But can't you get rid of the ranges entirely by inlining the regular expression check into the first loop? Just replace the push with the check?

Also, extract the code.slice out of the loop (even though we expect there to be only one).

Reply:
Done